### Лабораторная работа №3 (Варинат 7)
#### Выполнил студент группы 9383 Крейсманн Кирилл

### Цель:
Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

### Задачи:
- Решить свой вариант диофантова уравнения
- Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
- Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.




### Решение диофантова уравнения
$$2392x - 2461y = 138$$


In [8]:
import sympy

sympy.solvers.diophantine(2392*sympy.symbols('x')-2461*sympy.symbols('y')-138)

{(107*t_0 - 216, 104*t_0 - 210)}

$$x=-216+107*t$$
$$y=-210+104*t$$
$$t \in \mathbb{Z}$$


### RSA
Русское слово из четырех букв закодировано при помощи алгоритма ESA открытым ключом (e=15,m=46). Шифрованное сообщение имеет вид (36,7,43,34). Подберите закрытую часть ключа и прочитайте исходное слово. Буквам русского алфавита соответствуют числа в диапазоне от 2 до 32 (исклчены буквы "ё" и "ъ").


In [74]:
import sympy.crypto.crypto as cc
pk = cc.rsa_private_key(46,15)
print('Закрытая часть:',pk)
word=[36,7,43,34]
rs=[]

for i in word:
    print(cc.decipher_rsa(i,pk),end=' ')
    rs.append(cc.decipher_rsa(i,pk))
print('\n')




Закрытая часть: (46, 3)
12 21 19 20 



Получили числа - 12 21 19 20

Теперь получим слово:

In [51]:
for i in rs:
    print(chr(a+i-2),end='')

print('\n')
for i in range(2,33):
    print(i,end=' ')
print()

a=ord('а')
j=2
for i in range(a,a+32):
    if(chr(i)!='ъ'):
        print(chr(i),end=' ')
        if(j>=10):
            print('',end=' ')
        j+=1


куст

2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 
а б в г д е ж з и  й  к  л  м  н  о  п  р  с  т  у  ф  х  ц  ч  ш  щ  ы  ь  э  ю  я  

Получили слово "куст".

### ДНФ, КНФ

$$f(x,y,z) = ((x \vee y)\wedge(y \wedge x)) \oplus ((z \wedge x) \oplus (x \vee z))$$

In [71]:
import sympy.logic.boolalg as balg
x=sympy.symbols('x')
y=sympy.symbols('y')
z=sympy.symbols('z')

d=((x|y)&(y&x))^((z&x)^(x|z))

print(balg.to_dnf(d,True))
print(balg.to_cnf(d,True))

(y & z) | (z & ~x) | (x & ~y & ~z)
(x | z) & (z | ~y) & (y | ~x | ~z)


Получили,

ДНФ:
$$(y \wedge z) \vee (z \wedge \overline{x}) \vee (x \wedge \overline{x} \wedge \overline{z})$$

КНФ:

$$(x \vee z) \wedge (z \vee \overline{y}) \wedge (y \vee \overline{x} \vee \overline{z})$$

### Вывод:

В ходе выполнения данной лабораторной работы была изучена библиотека символьных вычислений Python (SymPy). С помощью данной библиотеки были решены три задачи.